numpyro/contrib/control_flow/scan.py

```diff
@@ -0,0 +1,196 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from functools import partial
+
+from jax import lax, random, tree_flatten
+import jax.numpy as jnp
+from jax.tree_util import register_pytree_node_class
+
+from numpyro import handlers
+from numpyro.primitives import _PYRO_STACK, apply_stack
+
+
+@register_pytree_node_class
+class PytreeTrace:
+    def __init__(self, trace):
+        self.trace = trace
+
+    def tree_flatten(self):
+        trace, aux_trace = {}, {}
+        for name, site in self.trace.items():
+            if site['type'] in ['sample', 'deterministic']:
+                trace[name], aux_trace[name] = {}, {'_control_flow_done': True}
+                for key in site:
+                    if key in ['fn', 'args', 'value', 'intermediates']:
+                        trace[name][key] = site[key]
+                    # scanned sites have stop field because we trace them inside a block handler
+                    elif key != 'stop':
+                        aux_trace[name][key] = site[key]
+        return (trace,), aux_trace
+
+    @classmethod
+    def tree_unflatten(cls, aux_data, children):
+        trace, = children
+        for name, site in trace.items():
+            site.update(aux_data[name])
+        return cls(trace)
+
+
+def _subs_wrapper(subs_map, i, length, site):
+    value = None
+    if isinstance(subs_map, dict) and site['name'] in subs_map:
+        value = subs_map[site['name']]
+    elif callable(subs_map):
+        rng_key = site['kwargs'].get('rng_key')
+        subs_map = handlers.seed(subs_map, rng_seed=rng_key) if rng_key is not None else subs_map
+        value = subs_map(site)
+
+    if value is not None:
+        value_ndim = jnp.ndim(value)
+        sample_shape = site['kwargs']['sample_shape']
+        fn_ndim = len(sample_shape + site['fn'].shape())
+        if value_ndim == fn_ndim:
+            # this branch happens when substitute_fn is init_strategy,
+            # where we apply init_strategy to each element in the scanned series
+            return value
+        elif value_ndim == fn_ndim + 1:
```

Comment:
So the time dim is always taken to be the first dim, even in the presence of outermost plate dims? I think that should be fine, but could you add a note in the docstring/comment, if that's the case?

Reply:
You are right, time dim will be the first dim. I think that is covered in the note in the docs (`plate` cannot be used as a context for `scan`). We also raise an error if users do
```
with plate(...):
    scan(...)
```